test/e2e/tests/test_scenarios.py

```diff
@@ -37,6 +37,88 @@ class ParameterGroupHelper(helper.ResourceHelper):
     Overrides methods as required for custom resources.
     """
 
+    def assert_items_k8s(self, expectations: dict, state: dict) -> None:
```

Comment:
Rather than overwriting the entire `assert_items_k8s` method, we should use the common logic that is defined for it in `ResourceHelper`. Then, when it gets to the custom logic inside the `for` loop, it could instead call another method which we overwrite. For example:

In `helper.py`:
```python
    def assert_items_k8s(self, expectations: dict, actual: dict) -> None:
        """Asserts the values in the `expectations` dict match the values as
        provided in the actual `dict`. This method calls into `assert_extra_k8s`
        for any items that can't be asserted in general.

        Args:
            expectations: dictionary with items (expected) to assert
            actual: dictionary of the latest spec or state of the K8s resource
        Returns:
            None
        """

        if not expectations:
            # nothing to assert as there are no expectations
            return

        if not actual:
            # there are expectations but no given state to validate
            # following assert will fail and assert introspection will
            # provide useful information for debugging
            assert expectations == actual

        for (key, value) in expectations.items():
            # conditions are asserted separately
            if key == "conditions":
                continue

            # tags are processed by custom logic
            if key == "tags":
                assertResourceTags(value, actual.get(key))
                continue

            self.assert_extra_k8s(key, value, actual)

            assert value == actual.get(key)

    def assert_extra_k8s(self, key: str, value: dict, actual: dict):
        # This method should be overridden by the any implementations of
        # the `ResourceHelper` that need to assert additional custom properties
        pass

    def assertResourceTags(self, expectedTags: dict, actualTags: dict):
        for expectedTag in expectedTags:
            tagKeyExist = False
            tagKey = expectedTag.get("key")

            for actualTag in actualTags:
                if ("key", tagKey) in actualTag.items():
                    assert expectedTag == actualTag
                    tagKeyExist = True
                    break

            if tagKeyExist:
                continue
```

Then in your individual implementations of `ResourceHelper`, extend it only the custom logic:
```python
    def assert_extra_k8s(self, key: str, value: dict, actual: dict):
        # This method should be overridden by the any implementations of
        # the `ResourceHelper` that need to assert additional custom properties
        if key == "parameterNameValues":
            if not value:
                assert value == actual.get(key)
                return

            for expectedParameterNameValue in value:
                parameterNameExist = False
                parameterName = expectedParameterNameValue.get("parameterName")
                latestParameterNameValues = actual.get(key)
                for latestParameterNameValue in latestParameterNameValues:
                    if ("parameterName", parameterName) in latestParameterNameValue.items():
                        assert expectedParameterNameValue == latestParameterNameValue
                        parameterNameExist = True
                        break
                if parameterNameExist:
                    continue
                assert expectedParameterNameValue == None
                
            return
```

Reply:
You should follow this same pattern for the boto3 style. It should greatly reduce code repetition and improve readability

Reply:
This way will miss a `continue` after calling `assert_extra_k8s`. I moved them to helper.py

Reply:
Created a function to override and pass boolean value to `assert_items_k8s`